distinguish implementation of data() and mutable_data() on TensorImpl

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98781
* #98780
* #98756
* #98755
* #98754
* #98753
* #98752
* #98751
* #98750
* #98734
* #98162
* #98163
* #97874
* #98720
* #98719
* __->__ #98732

The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

Commit:
distinguish implementation of data() and mutable_data() on TensorImpl

The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1564,31 +1564,55 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
  private:
-  // Templated implementation of data() and mutable_data().
+  // Shared implementation of data() and mutable_data() regardless of
+  // constness.
```
is the template really worth dealing with the TORCH_CHECK. Couldn't you have just factored the TORCH_CHECKs into a helper method

dagitses replied:
Up to you. But there's still the pointer math duplication.

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

ezyang reviewed (approved):
Much better!

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]

Commit:
Update on "distinguish implementation of data() and mutable_data() on TensorImpl"


The old style had them both going through a mutable method on Storage,
which would prevent us from implementing checks differently depending
on whether we are writing or reading.

Differential Revision: [D44831044](https://our.internmc.facebook.com/intern/diff/D44831044/)

[ghstack-poisoned]
